account_bank_statement_import_online/models/online_bank_statement_provider.py

```diff
@@ -0,0 +1,331 @@
+        ),
+        (
+            'valid_interval_number',
+            'CHECK(interval_number > 0)',
+            'Scheduled update interval must be greater than zero!'
+        )
+    ]
+
+    @api.model
+    def _get_available_services(self):
+        """Hook for extension"""
+        return []
+
+    @api.model
+    def _get_available_services_technical(self):
+        return self._get_available_services() + [('dummy', 'Dummy')]
+
+    @api.multi
+    @api.depends('service')
+    def _compute_name(self):
+        for provider in self:
+            provider.name = list(filter(
+                lambda x: x[0] == provider.service,
+                self._get_available_services_technical()
+            ))[0][1]
+
+    @api.multi
+    @api.depends('active', 'interval_type', 'interval_number')
+    def _compute_update_schedule(self):
+        for provider in self:
+            if not provider.active:
+                provider.update_schedule = _('Inactive')
+                continue
+
+            provider.update_schedule = _('%(number)s %(type)s') % {
+                'number': provider.interval_number,
+                'type': list(filter(
+                    lambda x: x[0] == provider.interval_type,
+                    self._fields['interval_type'].selection
+                ))[0][1],
+            }
+
+    @api.multi
+    def _pull(self, date_since, date_until):
+        AccountBankStatement = self.env['account.bank.statement']
+        if self.env.context.get('scheduled'):
+            AccountBankStatement = AccountBankStatement.with_context(
+                tracking_disable=True,
+            )
+        AccountBankStatementLine = self.env['account.bank.statement.line']
+        for provider in self:
+            statement_date_since = provider._get_statement_date_since(
+                date_since
+            )
+            while statement_date_since < date_until:
+                statement_date_until = (
+                    statement_date_since + provider._get_statement_date_step()
+                )
+                try:
+                    data = provider._obtain_statement_data(
+                        statement_date_since,
+                        statement_date_until
+                    )
+                except Exception as e:
+                    _logger.warning(
+                        'Online Bank Statement Provider failure: %s' % e
+                    )
+                    provider.message_post(
+                        body=str(e),
+                        subject=_('Online Bank Statement Provider failure'),
+                    )
+                    break
+                statement_date = provider._get_statement_date(
+                    statement_date_since,
+                    statement_date_until,
+                )
+                if not data:
+                    statement_date_since = statement_date_until
+                    continue
+                lines_data, statement_values = data
```

Comment:
Maybe, this could be a list of tuples. I mean, sometimes you receive the data from multiple bank accounts on the same call. For example, if you ask the bank all the information of your accounts he could give you all the related bank accounts of your user, that could be several for company or even, all the companies at the same time.

Reply:
But it is expected to receive data for one specific journal only, thus making this a list would over-complicate the code IMHO

Reply:
It might be harder, but this can happen (at least is what we found on our companies :cry: )

Reply:
In case you make overhead calls to banking API and filtering - how much the impact is? (In timing, or in cost per API call)? And I still don't have a good idea about how to enhance, taking into account that provider is attached to a journal, and thus having one provider for multiple journals would make this too complex to maintain

Reply:
It might be true, but sometimes they do not allow this. It is just my experience with some banks.

Reply:
However, this is not blocking, but it was interesting to consider.

Reply:
it would be great to hear out some more details on these specific banks, see API reference etc.

Reply:
As from what I've seen so far, everything is account-centric or at least holder-centric